File: ios/Classes/Flutter/ArcgisNativeObjectsController.swift
//
// Created by Valentin Grigorean on 25.07.2022.
//

import Foundation

protocol NativeMessageSink {
    func send(method: String, arguments: Any?) -> Void
}

protocol ArcgisNativeObjectFactory {
    func createNativeObject(objectId: Int, type: String, arguments: Any?, messageSink: NativeMessageSink) -> ArcgisNativeObjectController
}

class ArcgisNativeObjectsController: NativeMessageSink {
    private let channel: FlutterMethodChannel
    private let factory: ArcgisNativeObjectFactory
    private var nativeObjects: [Int: ArcgisNativeObjectController] = [:]

    init(messenger: FlutterBinaryMessenger, factory: ArcgisNativeObjectFactory) {
        channel = FlutterMethodChannel(name: "plugins.flutter.io/arcgis_channel/native_objects", binaryMessenger: messenger)
        self.factory = factory
        channel.setMethodCallHandler(handle)
    }

    func send(method: String, arguments: Any?) {
        channel.invokeMethod(method, arguments: arguments)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch (call.method) {
        case "createNativeObject":
            let args = call.arguments as! [String: Any]
            let objectId = args["objectId"] as! Int
            let type = args["type"] as! String
            let arguments = args["arguments"]
            let nativeObject = factory.createNativeObject(objectId: objectId, type: type, arguments: arguments, messageSink: self)
            nativeObjects[objectId] = nativeObject
            result(nil)
            break;
        case "destroyNativeObject":
            let objectId = call.arguments as! Int
            let nativeObject = nativeObjects.removeValue(forKey: objectId)
            nativeObject?.dispose()
            result(nil)
            break;
            case "sendMessage":
                let args = call.arguments as! [String: Any]
                let objectId = args["objectId"] as! Int
                let method = args["method"] as! String
                let arguments = args["arguments"]
                let nativeObject = nativeObjects[objectId]
                if let nativeObject = nativeObject {
                    nativeObject.onMethodCall(method: method, arguments: arguments, result: result)
                } else {
                    result(FlutterError(code: "object_not_found", message: "Native object not found", details: nil))
                }
                break;
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }

}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/flutter/ArcgisNativeObjectsController.java
package com.valentingrigorean.arcgis_maps_flutter.flutter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class ArcgisNativeObjectsController implements MethodChannel.MethodCallHandler, NativeMessageSink {
    private final MethodChannel channel;
    private final ArcgisNativeObjectFactory factory;
    private Map<Integer, ArcgisNativeObjectController> nativeObjects = new HashMap<>();

    public ArcgisNativeObjectsController(BinaryMessenger messenger, ArcgisNativeObjectFactory factory) {
        this.channel = new MethodChannel(messenger, "plugins.flutter.io/arcgis_channel/native_objects");
        this.factory = factory;
    }

    @Override
    public void send(@NonNull String method, @Nullable Object args) {
        channel.invokeMethod(method, args);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "createNativeObject": {
                final Map<?, ?> args = call.arguments();
                final int objectId = (int) args.get("objectId");
                final String type = (String) args.get("type");
                final Object arguments = args.get("arguments");
                final ArcgisNativeObjectController nativeObject = factory.createNativeObject(objectId, type, arguments, this);
                nativeObject.setMessageSink(this);
                nativeObjects.put(objectId, nativeObject);
                result.success(null);
            }
            break;
            case "destroyNativeObject": {
                final int objectId = call.arguments();
                final ArcgisNativeObjectController nativeObject = nativeObjects.get(objectId);
                if (nativeObject != null) {
                    nativeObject.setMessageSink(null);
                    nativeObject.dispose();
                    nativeObjects.remove(objectId);
                }
                result.success(null);
            }
            break;
            case "sendMessage": {
                final Map<?, ?> args = call.arguments();
                final int objectId = (int) args.get("objectId");
                final String method = (String) args.get("method");
                final Object arguments = args.get("arguments");
                final ArcgisNativeObjectController nativeObject = nativeObjects.get(objectId);
                if (nativeObject != null) {
                    nativeObject.onMethodCall(method, arguments, result);
                } else {
                    result.error("object_not_found", "Native object not found", null);
                }
            }
            default:
                result.notImplemented();
                break;
        }
    }


}
